### PR TITLE
Coin control dialog for CT and RingCT

### LINKED ISSUE #111: 3.) Coin Control needs to be functional
Coin Control (doesn’t work, no calcs, nothing brought forward to send tab, no zerocoin)

if users can’t send basecoin to ringct, these will need to be controlled some how in coin control. 

the issues I see is that presently miners mine to basecoin address (mining addresses)

### PR DESCRIPTION
**Pull request by ncnight**
The code seems to work on face and updates values properly. I still have to test if using the dialog sends the correct coins.

(fixes #111)

### CODE REVIEW

**Comment by ncnight** [comment]
No longer WIP and doesn't segfault from my testing on testnet. Coins are sent through the coin dialog and the coin control system is able to see them.

File: src/qt/coincontroldialog.cpp [hunk 1]
```diff
@@ -480,6 +483,29 @@ void CoinControlDialog::updateLabels(WalletModel *model, QDialog* dialog)
         else nBytesInputs += 148;
     }
 
+    AnonWallet * pAnonWallet = model->wallet().getWalletPointer()->GetAnonWallet();
+    std::map<uint256, CTransactionRecord> mapTxRecords = pAnonWallet->mapRecords;
+    for (const auto& recordPair : mapTxRecords) {
+        uint256 txid = recordPair.first;
+        CTransactionRecord txRecord = recordPair.second;
+        const CWalletTx* pWalletTx = model->wallet().getWalletPointer()->GetWalletTx(txid);
```
Comment (presstab) [inline src/qt/coincontroldialog.cpp]: segfault if this wallettx does not exist:
```
Thread 1 "veil-qt" received signal SIGSEGV, Segmentation fault.
CWalletTx::GetTxTime (this=this@entry=0x0) at wallet/wallet.cpp:1827
1827	    int64_t n = nTimeSmart;
(gdb) backtrace
#0  CWalletTx::GetTxTime (this=this@entry=0x0) at wallet/wallet.cpp:1827
#1  0x00000000005cdab1 in CoinControlDialog::updateView (this=this@entry=0x7fffffffc950) at qt/coincontroldialog.cpp:815
#2  0x00000000005cf067 in CoinControlDialog::setModel (this=this@entry=0x7fffffffc950, _model=0x30e6b20) at qt/coincontroldialog.cpp:160
#3  0x0000000000580ce9 in SendCoinsDialog::coinControlButtonClicked (this=0x3187fa0) at qt/sendcoinsdialog.cpp:774
#4  0x00007ffff6796d2a in QMetaObject::activate(QObject*, int, int, void**) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007ffff73f52b2 in QAbstractButton::clicked(bool) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#6  0x00007ffff7157f44 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#7  0x00007ffff7159559 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#8  0x00007ffff71596d4 in QAbstractButton::mouseReleaseEvent(QMouseEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#9  0x00007ffff709dfc8 in QWidget::event(QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#10 0x00007ffff705b05c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#11 0x00007ffff7060c19 in QApplication::notify(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#12 0x00007ffff676838b in QCoreApplication::notifyInternal(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#13 0x00007ffff705fb32 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#14 0x00007ffff70b85bb in ?? () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#15 0x00007ffff70bab7b in ?? () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#16 0x00007ffff705b05c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#17 0x00007ffff7060516 in QApplication::notify(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#18 0x00007ffff676838b in QCoreApplication::notifyInternal(QObject*, QEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#19 0x00007ffff6aaa4e1 in QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#20 0x00007ffff6aac1a5 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#21 0x00007ffff6a8ff08 in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#22 0x00007fffeda8f200 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5XcbQpa.so.5
#23 0x00007ffff3466197 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#24 0x00007ffff34663f0 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#25 0x00007ffff346649c in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#26 0x00007ffff67be7cf in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#27 0x00007ffff6765b4a in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#28 0x00007ffff676dbec in QCoreApplication::exec() () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#29 0x000000000045fc59 in main (argc=4, argv=0x7fffffffe058) at qt/veil.cpp:726
```

File: src/qt/coincontroldialog.cpp [hunk 2]
```diff
@@ -717,6 +744,117 @@ void CoinControlDialog::updateView()
         }
     }
 
+    AnonWallet * pAnonWallet = model->wallet().getWalletPointer()->GetAnonWallet();
+    std::map<uint256, CTransactionRecord> mapTxRecords = pAnonWallet->mapRecords;
+    std::map<CStealthAddress, std::tuple<CCoinControlWidgetItem*, CAmount, int>> mapTreeContainers;
+    for (const auto& recordPair : mapTxRecords) {
+        uint256 txid = recordPair.first;
+        CTransactionRecord txRecord = recordPair.second;
+        const CWalletTx* pWalletTx = model->wallet().getWalletPointer()->GetWalletTx(txid);
```
Comment (presstab) [inline src/qt/coincontroldialog.cpp]: Segfault if wallet tx does not exist.

File: src/qt/coincontroldialog.cpp [hunk 3]
```diff
@@ -717,6 +744,117 @@ void CoinControlDialog::updateView()
...
+                    ui->treeWidget->addTopLevelItem(itemNewAddressContainer);
+
+                    itemNewAddressContainer->setFlags(flgTristate);
+                    itemNewAddressContainer->setCheckState(COLUMN_CHECKBOX, Qt::Unchecked);
+
+                    // address
+                    itemNewAddressContainer->setText(COLUMN_ADDRESS, QString::fromStdString(sxAddress.ToString()));
+
+                }
+                auto const& mapEntry = mapTreeContainers[sxAddress];
```
Comment (presstab) [inline src/qt/coincontroldialog.cpp]: segfault when treeview is clicked
```
Thread 1 "veil-qt" received signal SIGSEGV, Segmentation fault.
__memcmp_sse4_1 () at ../sysdeps/x86_64/multiarch/memcmp-sse4.S:842
842	../sysdeps/x86_64/multiarch/memcmp-sse4.S: No such file or directory.
(gdb) backtrace
#0  __memcmp_sse4_1 () at ../sysdeps/x86_64/multiarch/memcmp-sse4.S:842
#1  0x00000000005cdb8c in CStealthAddress::operator< (y=..., this=0x38e73c0) at ./veil/ringct/stealth.h:77
#2  std::less<CStealthAddress>::operator() (this=0x7fffffffb670, __y=..., __x=...) at /usr/include/c++/5/bits/stl_function.h:387
#3  std::_Rb_tree<CStealthAddress, std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> >, std::_Select1st<std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> > >, std::less<CStealthAddress>, std::allocator<std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> > > >::_M_lower_bound (__k=..., __y=0x7fffffffb678, __x=0x38e73a0, 
    this=0x7fffffffb670) at /usr/include/c++/5/bits/stl_tree.h:1628
#4  std::_Rb_tree<CStealthAddress, std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> >, std::_Select1st<std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> > >, std::less<CStealthAddress>, std::allocator<std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> > > >::lower_bound (__k=..., this=0x7fffffffb670)
    at /usr/include/c++/5/bits/stl_tree.h:1091
#5  std::map<CStealthAddress, std::tuple<CCoinControlWidgetItem*, long, int>, std::less<CStealthAddress>, std::allocator<std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> > > >::lower_bound (__x=..., this=0x7fffffffb670) at /usr/include/c++/5/bits/stl_map.h:916
#6  std::map<CStealthAddress, std::tuple<CCoinControlWidgetItem*, long, int>, std::less<CStealthAddress>, std::allocator<std::pair<CStealthAddress const, std::tuple<CCoinControlWidgetItem*, long, int> > > >::operator[] (__k=..., this=0x7fffffffb670) at /usr/include/c++/5/bits/stl_map.h:479
#7  CoinControlDialog::updateView (this=0x7fffffffc950) at qt/coincontroldialog.cpp:792
```

**Pull request merged**